common/descriptor_checksum.c

```diff
@@ -0,0 +1,84 @@
+#include <ccan/short_types/short_types.h>
+#include <common/descriptor_checksum.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+
+static const char CHECKSUM_CHARSET[] = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+
+static const char INPUT_CHARSET[] =
+	"0123456789()[],'/*abcdefgh@:$%{}"
+	"IJKLMNOPQRSTUVWXYZ&+-.;<=>?!^_|~"
+	"ijklmnopqrstuvwxyzABCDEFGH`#\"\\ ";
+
+static inline int charset_find(char ch) {
+	for (size_t i = 0; i < sizeof(INPUT_CHARSET); i++) {
+		if (INPUT_CHARSET[i] == ch)
+			return i;
+	}
+	return -1;
+}
+
+static u64 polymod(u64 c, int val)
+{
+	u8 c0 = c >> 35;
+	c = ((c & 0x7ffffffff) << 5) ^ val;
+	if (c0 & 1) c ^= 0xf5dee51989;
+	if (c0 & 2) c ^= 0xa9fdca3312;
+	if (c0 & 4) c ^= 0x1bab10e32d;
+	if (c0 & 8) c ^= 0x3706b1677a;
+	if (c0 & 16) c ^= 0x644d626ffd;
+	return c;
+}
+
+
+int descriptor_checksum(const char *descriptor, int desc_size, char *checksum,
+			int checksum_capacity)
```

Comment:
Further, since this only ever returns `0` or `8`, you could update this to return a boolean.

Reply:
> and remove the usage of the 0 as a null terminator in this function

Argh actually it's a terrible idea: the checksum *is* a string.

> The Right Way to do this would be to introduce a new struct desc_checksum e.g.
struct desc_checksum {
        u8 csum[8];
}

Actually the right way is using `char`s